mongo-store.js

```diff
@@ -181,40 +185,163 @@ module.exports = function (opts) {
+
+
+                const NO_SPECIFIC_ID_REQUESTED = {}
+
+                const id = tryMakeIdIfRequested(args, {
+                  when_no_specific_id_requested: NO_SPECIFIC_ID_REQUESTED
+                })
+
+
+                if (id !== NO_SPECIFIC_ID_REQUESTED) {
+                  o.$setOnInsert = { _id: id }
+                }
+
+                return o
+              })()
+
+              return getcoll(args, args.ent, function (err, coll) {
+                if (error(args, err, cb)) {
+                  return
+                }
+
+                return coll.findOneAndUpdate(
+                  filter_by,
+                  replacement,
+                  { upsert: true, returnNewDocument: true },
+
+                  function (err, entu) {
+                    if (error(args, err, cb)) {
+                      return
+                    }
+
+                    return cb(null, { upsert_requested: true, out: entu })
+                  }
+                )
+              })
+            }
+          }
+
+          return cb(null, { upsert_requested: false, out: null })
+        }
+
+
+        function createNew(args, cb) {
+          return getcoll(args, args.ent, function (err, coll) {
+            if (error(args, err, cb)) {
+              return
+            }
+
+
+            const new_doc = (function () {
+              const public_entdata = args.ent.data$(false)
+
+
+              const NO_SPECIFIC_ID_REQUESTED = {}
+
+              const id = tryMakeIdIfRequested(args, {
+                when_no_specific_id_requested: NO_SPECIFIC_ID_REQUESTED
+              })
+
+
+              const new_doc = Object.assign({}, public_entdata)
+
+              if (id !== NO_SPECIFIC_ID_REQUESTED) {
+                new_doc._id = id
+              }
+
+              return new_doc
+            })()
+
+
+            return coll.insertOne(new_doc, function (err, inserts) {
+              if (error(args, err, cb)) {
+                return
+              }
+
+              const entu = inserts.ops[0]
+              let fent = null
 
-          coll.insertOne(entp, function (err, inserts) {
-            if (!error(args, err, cb)) {
-              var entu = inserts.ops[0]
-              var fent = null
               if (entu) {
                 entu.id = idstr(entu._id)
                 delete entu._id
-                //fent = ent.make$(_.cloneDeep(entu))
-                fent = ent.make$(seneca.util.deep(entu))
+                //fent = args.ent.make$(_.cloneDeep(entu))
+                fent = args.ent.make$(seneca.util.deep(entu))
               }
-              seneca.log.debug('save/insert', ent, desc)
-              cb(null, fent)
-            }
+
+              seneca.log.debug('save/insert', args.ent, desc)
+
+              return cb(null, fent)
+            })
           })
-        } else {
+        }
+
+
+        function tryMakeIdIfRequested(args, { when_no_specific_id_requested }) {
```

Comment:
OK - strong feedback here - please do not use 'if', 'and', etc in function names. Use the language constructs, and try to keep functions as pure as possible.

Use a `make_id` utility function in this case.